File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys
from collections import deque

sin = sys.stdin.readline
N = int(sin())
nums = list(map(int, sin().strip().split()))
ops = list(map(int, sin().strip().split()))
op_list = ["+", "-", "*", "/"]
max_val, min_val = -(sys.maxsize-1), sys.maxsize


def dfs(depth: int, arr):
    global max_val, min_val
    if depth == N:
        max_val = max(max_val, arr)
        min_val = min(min_val, arr)

    else:
        if ops[0] > 0:
            ops[0] -= 1
            dfs(depth+1, arr + nums[depth])
            ops[0] += 1

        if ops[1] > 0:
            ops[1] -= 1
            dfs(depth+1, arr - nums[depth])
            ops[1] += 1

        if ops[2] > 0:
            ops[2] -= 1
            dfs(depth+1, arr * nums[depth])
            ops[2] += 1

        if ops[3] > 0:
            ops[3] -= 1
            dfs(depth+1, int(arr / nums[depth]))
            ops[3] += 1


dfs(1, nums[0])

print(max_val)
print(min_val)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.*;

public class Main {
    public static int N, M;
    public static int min_val = Integer.MAX_VALUE;
    public static int max_val = Integer.MIN_VALUE;
    public static int[] opernands = new int [4];
    public static int[] numbers;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        String[] ipts;

        N = Integer.parseInt(br.readLine());
        numbers = new int[N];
        ipts = br.readLine().split(" ");

        for(int i = 0; i < N; i++){
            numbers[i] = Integer.parseInt(ipts[i]);
        }

        ipts = br.readLine().split(" ");

        for(int i = 0; i < 4; i++){
            opernands[i] = Integer.parseInt(ipts[i]);
        }

        dfs(1, numbers[0]);

        System.out.println(max_val);
        System.out.println(min_val);
    }

    public static void dfs(int depth, int cur){
        if(depth == N){
            min_val = Math.min(cur, min_val);
            max_val = Math.max(cur, max_val);
            return;
        }

        if(opernands[0] > 0){
            opernands[0]--;
            dfs(depth+1, cur+numbers[depth]);
            opernands[0]++;
        }

        if(opernands[1] > 0){
            opernands[1]--;
            dfs(depth+1, cur-numbers[depth]);
            opernands[1]++;
        }

        if(opernands[2] > 0){
            opernands[2]--;
            dfs(depth+1, cur*numbers[depth]);
            opernands[2]++;
        }

        if(opernands[3] > 0){
            opernands[3]--;
            dfs(depth+1, cur/numbers[depth]);
            opernands[3]++;
        }
    }
}
